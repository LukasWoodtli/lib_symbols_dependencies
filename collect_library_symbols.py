import locale
import os
import subprocess
import pathlib
from sys import platform

import utils
from dso import Dso


@utils.cache_result_to_file("dso_symbols")
def get_all_symbols_for_all_dsos(lib_dir):
    dsos = find_dsos(lib_dir)
    symbols = [get_all_symbols(dso) for dso in dsos]
    return symbols


def get_all_symbols(dso_path):
    print("Processing ", dso_path)
    defined = get_defined_symbols(dso_path)
    undefined = get_undefined_symbols(dso_path)
    return Dso(dso_path, defined, undefined)


FILE_SUFFIX = ".dylib" if platform == "darwin" else ".so"


def find_dsos(path):
    all_dsos = []
    for root, _directories, filenames in os.walk(path):
        for filename in filenames:
            file_path = pathlib.Path(root, filename)
            if file_path.suffix == FILE_SUFFIX:
                all_dsos.append(file_path)
    return set(all_dsos)


def get_undefined_symbols(so_file):
    ret = subprocess.run(["nm",
                          "--undefined-only",
                          "-C",
                          "-f", "posix",
                          so_file],
                         encoding=locale.getdefaultlocale()[1],
                         stdout=subprocess.PIPE,
                         check=True)
    symbols = []
    output = ret.stdout
    output = output.splitlines()
    for line in output:
        print(line)
        fields = line.split()
        assert len(fields) >= 2, fields
        symbol_type = fields[0].strip()
        name = fields[1].strip()
        if symbol_type == "U" and is_useful_symbol(name):
            symbols.append(name)
    return symbols


def get_defined_symbols(so_file):
    ret = subprocess.run(["nm",
                          "--defined-only",
                          "-C",
                          "--extern-only",
                          so_file],
                         encoding=locale.getdefaultlocale()[1],
                         stdout=subprocess.PIPE,
                         check=True)
    symbols = []
    for line in ret.stdout.splitlines():
        fields = line.split()
        assert len(fields) >= 3, fields
        symbol_type = fields[1].strip()
        name = fields[2].strip()
        if symbol_type == "T" and is_useful_symbol(name):
            symbols.append(name)
    return symbols


def is_useful_symbol(sym):
    unused_prefixes = ["_", "std::"]
    for prefix in unused_prefixes:
        if sym.startswith(prefix):
            return False
    charactes_in_unused_symbols = ["@"]
    for char in charactes_in_unused_symbols:
        if char in sym:
            return False
    return True
