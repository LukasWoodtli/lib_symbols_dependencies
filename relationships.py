import collections

import utils
from dso import Dso

RelationShip = collections.namedtuple("RelationShip", ["from_dso", "to_dso", "symbols"])
RelationShipStatistic = collections.namedtuple("RelationShipStatistic",
                                               ["dso",
                                                "outgoing_connection",
                                                "incoming_connections"])


class DsoRelationshipMatrix:

    def __init__(self):
        self._all_dsos = []
        self._relationship_data = collections.defaultdict(collections.OrderedDict)

    def add_dso(self, dso: Dso):
        for dso_already_in_data in self._all_dsos:
            self._insert_symbols_for_dsos(dso, dso_already_in_data)
            self._insert_symbols_for_dsos(dso_already_in_data, dso)

        self._relationship_data[dso.get_so_name()][dso.get_so_name()] = {}

        self._all_dsos.append(dso)

    def _insert_symbols_for_dsos(self, from_dso, to_dso):
        from_dso_name = from_dso.get_so_name()
        to_dso_name = to_dso.get_so_name()
        common_symbols = self._get_common_symbols(from_dso, to_dso)

        self._relationship_data[from_dso_name][to_dso_name] = common_symbols

    @staticmethod
    def _get_common_symbols(from_dso, to_dso):
        new_undefined = from_dso.get_undefined_symbols()
        defined = to_dso.get_defined_symbols()
        common_symbols = new_undefined & defined
        return common_symbols

    def get_relations(self, from_, to):
        return self._relationship_data[from_][to]

    def get_all_relations(self):
        rels = []
        for dso_from in self._all_dsos:
            for dso_to in self._all_dsos:
                dso_from_name = dso_from.get_so_name()
                dso_to_name = dso_to.get_so_name()
                if dso_from_name != dso_to_name:
                    symbols = self._relationship_data[dso_from_name][dso_to_name]
                    if len(symbols) > 0:
                        rels.append(RelationShip(dso_from_name,
                                                 dso_to_name,
                                                 symbols))
        return rels

    def get_all_relations_for_dso(self, dso):
        rels = self.get_relations_to_dso(dso)
        rels.extend(self.get_relations_from_dso(dso))

        rels.sort()
        return set(rels)

    def get_relations_from_dso(self, dso):
        rels = []
        for to_dso, symbols in self._relationship_data[dso].items():
            if to_dso != dso:
                if len(symbols) > 0:
                    rels.append(RelationShip(dso,
                                             to_dso,
                                             symbols))
        return rels

    def get_relations_to_dso(self, dso):
        rels = []
        for from_dso, items in self._relationship_data.items():
            if from_dso != dso:
                symbols = items[dso]
                if len(symbols) > 0:
                    rels.append(RelationShip(from_dso,
                                             dso,
                                             symbols))
        return rels

    def get_relations_for_every_single_dso(self):
        rels = collections.OrderedDict()
        for dso in self._all_dsos:
            name = dso.get_so_name()
            rels[name] = self.get_all_relations_for_dso(name)
        return rels

    def get_relations_stats_for_dso(self, dso):
        rels_to_dso = self.get_relations_to_dso(dso)
        num_rels_to_dso = sum([len(rel.symbols) for rel in rels_to_dso])

        rels_from_dso = self.get_relations_from_dso(dso)
        num_rels_from_dso = sum([len(rel.symbols) for rel in rels_from_dso])

        return RelationShipStatistic(dso, num_rels_from_dso, num_rels_to_dso)

    def get_relations_stats_for_all_dsos(self):
        return [self.get_relations_stats_for_dso(dso.get_so_name()) for dso in self._all_dsos]


@utils.cache_result_to_file("dso_relationships")
def construct_relationships(dsos):
    relationships = DsoRelationshipMatrix()
    for dso in dsos:
        relationships.add_dso(dso)
    return relationships
