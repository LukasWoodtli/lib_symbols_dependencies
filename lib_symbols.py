#!/usr/bin/env python3
import utils
from collect_library_symbols import get_all_symbols_for_all_dsos
from relationships import construct_relationships
from render_graphs import create_relationship_graphs
from save_statistics import save_as_csv


def collect_symbols_and_create_graphs(lib_dir, output_dir):
    dsos = get_all_symbols_for_all_dsos(lib_dir)
    relationships = construct_relationships(dsos)

    create_relationship_graphs(relationships, output_dir)

    save_as_csv(relationships, ".")


if __name__ == "__main__":
    utils.ENABLE_CACHING = True
    lib_dir = "lib"

    collect_symbols_and_create_graphs(lib_dir, "graphs")
