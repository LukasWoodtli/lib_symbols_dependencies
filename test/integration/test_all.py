import collections
import json
import os
import shutil
import subprocess
import unittest
from sys import platform

import approvaltests
import pytest

import collect_library_symbols
from lib_symbols import collect_symbols_and_create_graphs
from relationships import construct_relationships
from collect_library_symbols import get_all_symbols_for_all_dsos

SCRIPT_DIR = os.path.split(__file__)[0]
EXAMPLES_DIR = os.path.join(SCRIPT_DIR, "example_libraries")
EXAMPLES_BUILDDIR = os.path.join(EXAMPLES_DIR, "build")
EXAMPLES_DSOS = os.path.join(EXAMPLES_BUILDDIR, "lib")

GRAPH_OUTPUT_PATH = os.path.join(SCRIPT_DIR, "graphs")


def build_example_dsos():
    if os.path.isdir(EXAMPLES_BUILDDIR):
        shutil.rmtree(EXAMPLES_BUILDDIR)
    os.mkdir(EXAMPLES_BUILDDIR)
    cmake_command = f"cmake {EXAMPLES_DIR}"
    subprocess.run(cmake_command.split(), cwd=EXAMPLES_BUILDDIR)
    subprocess.run("make", cwd=EXAMPLES_BUILDDIR)


@pytest.fixture(autouse=True)
def set_up():
    build_example_dsos()

    if os.path.isdir(GRAPH_OUTPUT_PATH):
        shutil.rmtree(GRAPH_OUTPUT_PATH)

    collect_symbols_and_create_graphs(EXAMPLES_DSOS, GRAPH_OUTPUT_PATH)


def _get_file_path(graph_name):
    file_name = f"dsos.{graph_name}.neato.gv"
    graph_path = os.path.join(GRAPH_OUTPUT_PATH, file_name)
    return graph_path


def test_find_dsos():

    expected = set(['libbaz', 'libfoo', 'libbar'])
    if platform == "darwin":
        expected = [expected + ".dylib" in expected]
    else:
        expected = [expected + ".so" in expected]

    so_files = collect_library_symbols.find_dsos(EXAMPLES_DSOS)

    so_files = [os.path.basename(file) for file in so_files]
    so_files = set(so_files)
    assert expected == so_files


@pytest.mark.parametrize("test_library,defined_symbols",
                         [("libfoo.so", ["foo", "foo2"]),
                          ("libbar.so", ["bar"]),
                          ("libbaz.so", ["baz"])])
def test_get_defined_symbols(test_library, defined_symbols):
    symbol_list = collect_library_symbols.get_defined_symbols(os.path.join(EXAMPLES_DSOS, test_library))

    assert set(symbol_list) == set(defined_symbols)


@pytest.mark.parametrize("test_library,undefined_symbols",
                         [("libfoo.so", ["bar"]),
                          ("libbar.so", ["foo2"]),
                          ("libbaz.so", ["foo2"])])
def test_get_undefined_symbols(test_library, undefined_symbols):
    symbol_list = collect_library_symbols.get_undefined_symbols(os.path.join(EXAMPLES_DSOS, test_library))

    assert set(symbol_list) == set(undefined_symbols)


def test_all():
    approvaltests.verify_file(_get_file_path("all"))


def test_libbar():
    approvaltests.verify_file(_get_file_path("libbar.so"))


def test_libbaz():
    approvaltests.verify_file(_get_file_path("libbaz.so"))


def test_libfoo():
    approvaltests.verify_file(_get_file_path("libfoo.so"))


def test_relationships():
    dsos = get_all_symbols_for_all_dsos(EXAMPLES_DSOS)
    relationships = construct_relationships(dsos)

    all_rels = collections.OrderedDict()
    for rel in relationships.get_all_relations():
        all_rels[f"{rel.from_dso} -> {rel.to_dso}"] = len(rel.symbols)

    approvaltests.verify(json.dumps(all_rels, indent=2))
