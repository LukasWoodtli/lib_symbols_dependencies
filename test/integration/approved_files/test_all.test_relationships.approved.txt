{
  "libbar.so -> libfoo.so": 1,
  "libbaz.so -> libfoo.so": 1,
  "libfoo.so -> libbar.so": 1
}
